### DISCUSSION TITLE
HomeMatic device names / HA entity names

### DISCUSSION THREAD

**Discussion by noxhirsch** [SukramJ/aiohomematic#92] [Q&A] [answered]
I'm a little confused about the device/entity names:
In HomeMatic this device, for example, is called "ws_charger". The device in HA is also named like this - as well as the diagnostic sensors. For normal sensors and controls, however, the name is not adopted. 
Is this not yet implemented or a bug? Or must there be something wrong in my installation? I have already deleted the integration and removed all remaining entries from the entity registry, but at least that doesn't seem to be the problem.

![image](https://user-images.githubusercontent.com/30938717/147359296-4f304e53-d40a-4ae1-b177-31c02cdd3210.png)

**Comment by SukramJ** [comment, marked as answer]
Did you set the name of channels in your CCU?

**Reply by noxhirsch** [reply]
No, you're right - that was the problem! Thank you for the fast help!

I guess a lot of users probably won't do that. Maybe a solution here would be to overwrite the channel name with the device name if the device type (in this case "HM-ES-PMSw1-Pl-DN-R1") is in it?

**Reply by SukramJ** [reply]
I recommend using [Homematic Manager](https://github.com/hobbyquaker/homematic-manager) to rename channels.
It's faster than using the CCU WebUI and there is an option to use the device name for channel name.

After renaming anything in CCU it's necessary to reload the integration in HA to clear the names cache. That would use the new names for entities/devices. If you also want to have new entity_ids it's necessary to delete and create the integration.

**Reply by noxhirsch** [reply]
Thank you very much - that helped a lot!

**Reply by emufan** [reply]
> I guess a lot of users probably won't do that. Maybe a solution here would be to overwrite the channel name with the device name if the device type (in this case "HM-ES-PMSw1-Pl-DN-R1") is in it?

@SukramJ I like this suggestion as well. Esp. as I'm not able to get the HM Manager running (endless loading) and I agree with noxhirsch, that only some of the user have named channels. Esp. those, who are using HA.

**Reply by Lemocin** [reply]
> I'm not able to get the HM Manager running (endless loading) and I agree with noxhirsch, that only some of the user have named channels. Esp. those, who are using HA.

I was able to fix the endless loading with the help of [this post](https://github.com/hobbyquaker/homematic-manager/issues/121#issuecomment-749725958). 
But i also agree, if i rename devices in other Integrations, HA prompts me to rename the entities automatically as well. 
I think this approach would be better than the current. Anyway i'm also fine with renaming.

**Reply by SukramJ** [reply]
So what's the rule to use the device_name instead of the channel_name?
Pseudocode:
Opt 1 `if device_type in channel_name then use device_name`
Opt 2  `if device_type and channel_address  in channel_name then use device_name`

Opt2 should be more precise if the channel name is untouched

**Reply by SukramJ** [reply]
How do you work with ccu programs and direct connections without naming the channels?

**Reply by emufan** [reply]
> How do you work with ccu programs and direct connections without naming the channels?

I don't have programs. I do everything in HA. And for direct connections I don't get the question, because the device name is listed in the new-dialog as well. That is enough for me to find the relevant one. 

Opt2 above would be fine. Should work for the situation, that the ones are untouched.

**Reply by Lemocin** [reply]
> How do you work with ccu programs and direct connections without naming the channels?

Well, i didn't know i can rename the channels. So i just "remembered" the addresses on my second screen. The webui is more of a backend for me and everything  that should be "fancy" occurs in HA. 

I agree that Opt 2 would be the best in my opinion.

**Reply by noxhirsch** [reply]
I'd vote for Opt2, too.

> How do you work with ccu programs and direct connections without naming the channels?

I only have a few heating groups - direct connections I have never touched myself and programs only very rarely. Since I have a pretty wild mix of devices (HM,Zigbee,ESPHome,...) at the moment, I automate everything in HomeAssistant.

**Reply by SukramJ** [reply]
Please test the latest release

**Reply by Lemocin** [reply]
> Please test the latest release

Seems to work. I don't really know why it only shows the 4 in some views.
![grafik](https://user-images.githubusercontent.com/4162496/147791347-7b6df0dd-4ffb-4533-8fab-3170095f8dbd.png)
Even if i rename the entity, It stay's on 4 on the autogenerated lovelace card. If i re-add the card then it changes. 
On the device view it discards always the Device name in the entity name but i think that is intended from HA(so it justs says e.g. Level and not Roomname Levelin the device view).

**Reply by SukramJ** [reply]
If you rename `Gästezimmer 4` to `Gästezimmer` it should be displayed on all cards.
if you change it in  the CCU you have to click on reload integration (integrations page) to clear the caches. A restart of HA doesn't clear the caches.

**Reply by Lemocin** [reply]
> If you rename `Gästezimmer 4` to `Gästezimmer` it should be displayed on all cards.

Yes, it's not a problem. Just the autogenerated card doesn't change it's name. But that's an issue from HA or Lovelace, not of this integration i think.

**Comment by emufan** [comment]
Can we perhaps discuss the entity namings again?

Old integration:
Main Entity name is set a device name in HM
Additional Entity names are set a device name in HM plus Suffix.
=> So the user is used to get the name from HM as main entity name. The entity name is used in the UIs.

Current state of new integration
Device name is set a device name in HM.
Every Entitiy name is set as device name in HM but with suffix of the channel

Proposal:
Device name is set as device name in HM.
**Main entitiy name is set as device name in HM as well**
Additional channels are set as device name in HM but with suffix of the channel
=> The user has still the name from HM as entity name. The entity name is used in the UIs. This is common for other integrations as well. Device name = Main entity name.
=> This is then more or less the same as in old integration (no breaking) and most usefull.

**Reply by SukramJ** [reply]
What are your arguments to change the current schema?
What is a main entity?
How do you identify a main entity?

**Reply by emufan** [reply]
My argument is, that if I use the entity in UI, currently it has always a not usable name. So nearly every name in UI display or entity settings would have to be renamed back to device name.

E.g.
Heizung Kellerflur 1
Rolladen Flur oben 4
Schreibtisch-Aktor 4
Terrassentür State
etc.

The status is, no one will use this names and have to or will rename all of them. And I promise, that this would be a major point, if this one wil replace the other one or more users will migrate,

The main entity? The one who got the HM name without a suffix in the old integration. 😉

**Reply by SukramJ** [reply]
Custom Entities like switch, cover, climate:
There is already a solution in place: Change the name of the channel in CCU! This is a one time action.
Rename channel 1 in CCU to Heizung Kellerflur and that ist. Afterwards reload the integration.

Auto-Generated Entities sensor, binary_sensor, number, select:
The main feature of this integration is that we don't want to maintain code to define devices/entities to make them usable in HA.
So we don't maintain a list of primary entities.

**Reply by emufan** [reply]
> There is already a solution in place: Change the name of the channel in CCU! This is a one time action.

Yes I know. But I'm prety sure, that that will bring a lot of complains in the future, if this will be the official successor.

> The main feature of this integration is that we don't want to maintain code to define devices/entities to make them usable in HA.

Hm, but you are already doing this for known device classes, don't you? And you are generating only one switch per BROLL, BSM, ... (which is the main entity, where I'm talking about I think) as before (and before this one got the HM name) and the others only when activation the "generate virtual channels".

**Reply by SukramJ** [reply]
There will always be complains, and there will always be different opinions. Therefore rules are important.
Once, someone is willing to contribute documentation, there will be a paragraph naming "If you want nicer names, name your channel"

Why are device_classes relevant?

**Reply by emufan** [reply]
I thought you are already distinguishing because of this, which channels are generated. E.g. channel 1 for thermostat or channel 4 for bsm or channel 4 for broll, ... Or how they are generated different to other classes. Or be able to distinguish, for which the next proposed rule could apply.

But why not using the rule: the once, which are getting generated in standard (without "generate virtual channels") are getting the device name. Voila. Simple, but rule. 😊 And could be used for climate, cover, switch and light.

**Reply by SukramJ** [reply]
What happens if a user activates the virtual channels (5+6)?
What happens if channel 5 is named like the device? How to handle duplicate names?
HA uses names for entity_id. If these are not unique HA appends a number.
Renaming the primary channel gives you the required uniqueness, because channel names must be unique within a device.
I know how to identify the primary channel of a **custom_entity**, but i don't think this is a good solution when thinking about uniqueness.

**Reply by emufan** [reply]
Hm. Don't see the difference to my proposed rule:

a. The once, which are getting generated in standard (**without "generate virtual channels"**) are getting the device name
b. All the other once, additional channels, ... as current naming implementation. 

As far as I can see: Uniqueness still there. 

But go ahead and leave it as it is. I only see the disadvantages, if I compare to the old integration and only never like deteriorations to the current state.

**Reply by SukramJ** [reply]
Feel free to create a PR. That might give you an insight of the complexity of naming things.
<del>All i could do is to cut off the channel number for the primary channel. </del>
[Edit]: it's not a real option and will break multi channel devices.

I've never thought that is so hard to fulfill some prerequisites like naming some channels for custom entities.

**Reply by SukramJ** [reply]
There is one option with limited effort:
If device is **not a multi device** (multiple physical channels) and virtual channels are **disabled**:
Cut the channel number from the entity name

If device is **a multi device** (multiple physical channels) or virtual channels are **enabled**:
**All channels** will habe a channel suffix, **unless the channel name  is customized in CCU**

**Reply by emufan** [reply]
Sounds great at least for me.

But think about Sunday lunch and Sunday relaxing. 😂

**Reply by danielperna84** [reply]
@SukramJ 
Regarding the option with limited effort: which effect will this have if a user decides to enable the virtual channels, but then decides to go back to disabled channels? Would this create inconsistent / broken entities that could cause problems if not properly cleaned up? Just mentioning this as a possible _problem_ that could be introduced.

**Reply by SukramJ** [reply]
I don‘t think so. The entity id will not change after initial setup, due to the unique id. Only the displayed entity name will change, if not customized in ha.
This is only relevant for hmip devices with one physical and virtal channels. So basically switches and light.
Climate has no virtual channels., and Cover displays no virtual channels as described in another discussion.
